controllers/component_dependency_update_controller.go

```diff
@@ -0,0 +1,375 @@
+		return r.removePipelineFinalizer(ctx, pipelineRun)
+	}
+	// find the image and digest we want to update to
+	image := ""
+	digest := ""
+	for _, r := range pipelineRun.Status.Results {
+		if r.Name == ImageDigestParamName {
+			digest = r.Value.StringVal
+		} else if r.Name == ImageUrlParamName {
+			image = r.Value.StringVal
+		}
+	}
+	// Failure to find is a permanent error, so remove the finalizer
+	if image == "" {
+		log.Error(fmt.Errorf("unable to find %s param on PipelineRun, not performing nudge", ImageUrlParamName), "no image url result")
+		return r.removePipelineFinalizer(ctx, pipelineRun)
+	}
+	if digest == "" {
+		log.Error(fmt.Errorf("unable to find %s param on PipelineRun, not performing nudge", ImageDigestParamName), "no image digest result")
+		return r.removePipelineFinalizer(ctx, pipelineRun)
+	}
+
+	alreadyProcessed := map[string]bool{}
+	if pipelineRun.Annotations != nil && pipelineRun.Annotations[AlreadyNudgedAnnotationName] != "" {
+		list := strings.Split(pipelineRun.Annotations[AlreadyNudgedAnnotationName], ",")
+		for _, i := range list {
+			alreadyProcessed[i] = true
+		}
+	}
+
+	components := applicationapi.ComponentList{}
+	err := r.Client.List(ctx, &components, client.InNamespace(pipelineRun.Namespace))
+	if err != nil {
+		log.Error(err, "failed to list components in namespace")
+		return ctrl.Result{}, err
+	}
+
+	retryTime := FailureRetryTime
+	immediateRetry := false
+	updateFailed := false
+	for i := range components.Items {
+		comp := components.Items[i]
+		if alreadyProcessed[comp.Name] {
+			continue
+		}
+		if comp.Spec.Application == updatedComponent.Spec.Application && slices.Contains(updatedComponent.Spec.BuildNudgesRef, comp.Name) {
+			var nudgeErr error
+			immediateRetry, nudgeErr = r.UpdateFunction(ctx, r.Client, r.EventRecorder, updatedComponent, &comp, image, digest)
+			if nudgeErr != nil {
+				log.Error(nudgeErr, fmt.Sprintf("component update of %s as a result of a build of %s failed", comp.Name, updatedComponent.Name), l.Audit, "true")
+				updateFailed = true
+				continue
+			} else {
+				alreadyProcessed[comp.Name] = true
+			}
+		}
+	}
+	if updateFailed {
+
+		if pipelineRun.Annotations == nil {
+			pipelineRun.Annotations = map[string]string{}
+		}
+		existing := pipelineRun.Annotations[FailureCountAnnotationName]
+		if existing == "" {
+			existing = "0"
+		}
+		failureCount, err := strconv.Atoi(existing)
+		if err != nil {
+			log.Error(err, "failed to pass retry count, not retrying")
+			return r.removePipelineFinalizer(ctx, pipelineRun)
+		}
+		failureCount = failureCount + 1
+
+		pipelineRun.Annotations[FailureCountAnnotationName] = strconv.Itoa(failureCount)
+
+		//update the list of successful updates
+		//if any
+		if len(alreadyProcessed) > 0 {
+			val := ""
+			for i := range alreadyProcessed {
+				if val != "" {
+					val = val + ","
+				}
+				val = val + i
+			}
+			pipelineRun.Annotations[AlreadyNudgedAnnotationName] = val
```

Comment:
What if a new build will start in meantime?

Reply:
That is what I mentioned above about builds going stale. When I perform the 'nudge' I could check for currently running pipelines for the same component that are older than the current state, and if they exist and are still running mark them as already nudged and remove the finalizer.

The other approach would be that when a pipeline finishes we check if a newer pipeline exists and has finished, however I don't think that approach will work very well as the pruner may have removed already completed pipelines.

Reply:
I have implemented the stale check logic